pkg/builder/controller.go

```diff
@@ -232,6 +249,12 @@ func (blder *TypedBuilder[request]) WithEventFilter(p predicate.Predicate) *Type
 	return blder
 }
 
+// WithHandlerWrapper sets an optional wrapper function that is applied to the TypedEventHandler for watches set up via the Watches func.
+func (blder *TypedBuilder[request]) WithHandlerWrapper(handlerWrapper func(handler.TypedEventHandler[client.Object, request], cluster.Cluster) handler.TypedEventHandler[client.Object, request]) *TypedBuilder[request] {
+	blder.handlerWrapper = handlerWrapper
+	return blder
+}
```

Comment:
A note on this: I would have preferred for this to be an option that can be passed to `Watches`, but the main challenge there was the `untypedWatchesInput` interface not allowing to have a generic type for `request` (since it's untyped). I'm still very much open to making it an opt, but I haven't found a good way to adjust the existing interface without breaking existing usage.